### ISSUE TITLE
bug: Batch Normalization moving average of variance is initialized as 0

### ISSUE THREAD

**Issue by pluskid** [deepmind/sonnet#174]
The moving average of the variance in Batch Normalization is initialized as 0 at [here](https://github.com/deepmind/sonnet/blob/v2/sonnet/src/batch_norm.py#L314) and [here](https://github.com/deepmind/sonnet/blob/v2/sonnet/src/moving_averages.py#L98). As a result, when I directly evaluate a randomly initialized network without training it first, the results are always NaN.

**Comment by tomhennigan** [comment]
Hey @pluskid I suspect this is only the case if you are setting `eps=0`, we (inline with the paper) include a small epsilon (default: `1e-5`) to avoid division by zero (https://sonnet.readthedocs.io/en/latest/api.html#batchnorm) and indeed for a randomly initialized batchnorm layer I do not get `NaN` for any value of is_training+test_local_stats:

```python
x = tf.random.normal([4, 4])
for is_training in (True, False):
  for test_local_stats in (True, False):
    bn = snt.BatchNorm(True, True)
    y = bn(x, is_training=is_training, test_local_stats=test_local_stats)
    assert not tf.reduce_any(tf.math.is_nan(y))
```

If you want to keep `eps=0` and avoid division by zero you could pass `test_local_stats=True` to your BatchNorm layers to use the mean/variance of the current batch rather than the average.

**Comment by pluskid** [comment]
@tomhennigan Thanks! I did not set the default epsilon. But I was using a large network (ResNet50), not sure if that is overflow caused by magnifying from many BatchNorm layers in the network. I will leave the issue closed, and will come back if I got a chance to make a minimum reproducible example.

**Comment by wookayin** [comment]
@tomhennigan Even if eps > 0 (say 1e-3), the first moving variance reported is regardless 0.0 which makes the activation too large. It will always blow up the layer and training as well -- in the very first training iteration loss can go infinite, because the moving variance would be updated AFTER the first forward training pass. So I don't think the current implementation of initialization is correct.

Please note that Keras BatchNorm initializes moving variance with one: https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/layers/normalization.py#L133

**Comment by wookayin** [comment]
I figured out that the root cause why I was experiencing a "zero moving variance" in the first training step is because there was a dummy call with` tf.zeros` tensors to ensure variables created (where `is_training` is set to `False`), as in [tf_utils.create_variables](https://github.com/deepmind/acme/blob/master/acme/tf/utils.py#L78-L95). However, the current implementation of `snt.BatchNorm` assumes that variables would be created with the first batch --- **regardless of `is_training=False`**. If the argument `is_training=False` was given, it should NOT initialize the variance, but it appears that the dummy calls would make the moving variance incorrectly initialized. If variable initialization is done *before* the first forward pass (e.g. initialize the moving average with 1), there would be no problem.
